### ISSUE TITLE
[Feature Request] Let file operations finish before automatically locking volume

### ISSUE THREAD

**Issue by creeper451** [hardcore-sushi/DroidFS#157]
The app stops importing files and crashes when i switch to something else, is this normal?
Or is it for security purposes because i know that you could enable the unsafe feature "Keep volume open when the app goes in background", i would prefer it to close after it finishes the importing.

**Comment by hardcore-sushi** [comment]
Yes, it's a security feature. Volumes are closed as soon as possible, whether or not a file transfer operation is in progress. Otherwise, in the event of a large file transfer or a bug that blocks the transfer, the volume could remain open for a very long time, which could be a security issue.

**Comment by creeper451** [comment]
@hardcore-sushi Ok, but can't we have an option in the unsafe feature section to close the volume after completing file transfer and make that if you enable it, you cannot enable "Keep volume open when the app goes in background", at the same time

**Comment by mattfbacon** [comment]
I would also like it to keep the device awake to avoid locking that way. It should probably just show a progress dialog. There could be an option to "Continue in background" which would explicitly allow this old behavior of ending the transfer whenever the app loses focus.

**Comment by hardcore-sushi** [comment]
OK I think this should be implemented as an unsafe feature. Something like "Let file operations finish before automatically locking volume"

**Comment by mattfbacon** [comment]
Or in general just some indication that an operation is in progress, within the app and/or as a progress notification

**Comment by CodeCubeNeo** [comment]
Or a warning (with cancel, ok) when you try to close the volume that on operation is still going on?

**Comment by hardcore-sushi** [comment]
> Or in general just some indication that an operation is in progress, within the app and/or as a progress notification

This won't prevent the volume to be locked if the app goes in background.

> Or a warning (with cancel, ok) when you try to close the volume that on operation is still going on?

I think you're not talking about the same thing. The original topic of this issue is delaying automatic locking when file operations are in progress and the app goes in background, not delaying volume lock triggered by the user within the app.

**Comment by mattfbacon** [comment]
Also, assuming that the application acts as a single entry point to access the data, this "operations in progress" mode could act as if the volume is locked and the user does need to provide a password but actually leave the volume unlocked in the background so the operation can continue.

**Comment by hardcore-sushi** [comment]
@mattfbacon Volume locking is not about preventing the user to open it, it's about wiping all cryptographic secrets from memory, so that we are not able to decrypt the volume even if we dumped storage and RAM. What you describe would give a false sense of security.

**Comment by Anon7250** [comment]
> Volume locking is not about preventing the user to open it, it's about wiping all cryptographic secrets from memory, so that we are not able to decrypt the volume even if we dumped storage and RAM. What you describe would give a false sense of security.

I've also been thinking about this. To prove that DroidFS is safe, we are assuming RAM is not being read by other processes.

But, I think, if someone on the phone, can just dump RAM. Let's say they have gained root permission, and can basically debug / trace / attack any program. Then, it's already unsafe, whether we keep volumes open or not. Anything read by the application is not safe, because we are doing decryption in the user space. 

Basically, today an assumed rooted attacker can already extract files when DroidFS is in the foreground. 

I think the goal of DroidFS is not to achieve perfect security, but just for the added encryption itself. DroidFS should not need to solve for RAM dumping. For powerusers, we need to harden our own devices. For normal users, we need to update our softwares to ensure OS integrity. 

So, I think we should allow volumes to stay open, temporarily in the background, to complete file operations. Does it make sense?

**Comment by hardcore-sushi** [comment]
@Anon7250 The automatic volume lock feature is not intended to protect against software compromise, but rather against physical memory dumps. This is useful, for example, if your device gets seized by surprise. In this threat model, it's essential to minimize the time during which secrets are exposed in memory.

**Comment by Anon7250** [comment]
Please elaborate on how this physical memory dump can be achieved without root access. As far as I understand, normally nobody has permissions to debug a production app (see [android:debuggable](https://developer.android.com/privacy-and-security/risks/android-debuggable)). 

OEM unlocking cannot be attempted without a data wipe. And if the device is already rooted, then it's up to the device owner (many root solutions allow limiting the permissions, such as preventing adb from gaining root)

**Comment by hardcore-sushi** [comment]
By physical memory dump, I mean disassembling the device and plugging directly to the motherboard to read the RAM physically, thus bypassing all software protections. This can also be done less brutally by using manufacturers' protocols to talk directly to the chips.

But for lower threat models, it could be not allowing someone to access the volume if the user has not had time to lock the screen or if the lock screen is weak (biometric authentication, simple PIN/pattern).

**Comment by Anon7250** [comment]
Got it. In that case, I think maybe we can add this background file
operation as an unsafe feature for users to toggle on or off.

Also, even if we toggle off and forbid background operations on files, we
should make sure the operations are merely paused and can be resumed when
the volume is decrypted again.

This sounds interesting and maybe I can try and help implement something at
the end of this year.

**Comment by hardcore-sushi** [comment]
> maybe we can add "keeping volumes temporarily open in background to complete file operations" as an unsafe feature for users to toggle on or off.

Yes, that's what's planned.

> we should make sure the operations are merely paused and can be resumed when the volume is decrypted again.

That would be great, but how could we store file paths safely while the volume is closed?

**Comment by Anon7250** [comment]
> how could we store file paths safely while the volume is closed?

If file paths alone are also considered sensitive information, then,

One solution is to store the list of pending operations inside the
encrypted volume itself.

We can store them as a json file, perhaps: `<encrypted
volume>/.droidfsInternals/pendingOperations.json`

Would that suffice?

**Comment by hardcore-sushi** [comment]
> One solution is to store the list of pending operations inside the encrypted volume itself.

Yes, this can work, but it's not always desirable, as users may synchronize their volumes between devices and therefore not want to modify only one of the copies. Also, exporting files can be done from read-only volumes, so it won't be possible to store data inside them.

Another idea would be to use a separate hidden volume created with the same encryption key as the real volume, so that both can be opened at the same time. However, this would introduce a lot of overhead.
